internal/operators/operators.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operators
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	columnhelpers "github.com/inspektor-gadget/inspektor-gadget/internal/column-helpers"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type Runner interface {
+	ID() string
+	Columns() columnhelpers.Columns
+	Gadget() gadgets.Gadget
+	Context() context.Context
+	Operators() Operators
+	Logger() logger.Logger
+}
+
+type Operator interface {
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// Params will return global params (required) for this operator
+	Params() params.Params
+
+	// PerGadgetParams will return params (required) per gadget instance of the operator
+	PerGadgetParams() params.Params
+
+	// Dependencies can list other operators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.Gadget) bool
+
+	// Init allows the operator to initialize itself
+	Init(params params.Params) error
+
+	// Cleanup allows the operator to clean up stuff prior to exiting
+	Cleanup() error
```

Comment:
What about using `Close() error` in all those functions to indicate a resource is tear down?